externallib/DPA/DP.py

```diff
@@ -0,0 +1,152 @@
+"""
```

Comment:
I'm not sure we want to commit the comments I've added to this class. This was me taking notes while attempting to understand the class rather than an attempt to document it myself!